File: packages/validator-rust/src/expr/parser.rs
//! CRUDUI expression parser: `Vec<Token>` → AST (expressions.md §3/§4, JS/PHP
//! `Parser` parity).
//!
//! Recursive-descent grammar (EBNF, lowest precedence first):
//!
//! ```text
//!   ternary    = or [ "?" ternary ":" ternary ]        (right-associative)
//!   or         = and { "||" and }
//!   and        = not { "&&" not }
//!   not        = "!" not | comparison
//!   comparison = primary [ cmp_op cmp_value | in_op value_list ]
//!   primary    = "(" or ")" | path | literal
//!   path       = [ "." | ".." ] identifier { "." (identifier | number | "*") }
//! ```
//!
//! Priority and right-associativity come from the grammar productions, not a
//! heuristic. No eval, no split (GRAMMAR §10).

use super::ast::{LiteralValue, Node, PathSegment};
use super::error::ParseError;
use super::token::{Literal, Token, TokenType};

/// The most nodes on a path from the root of an expression's syntax tree to a
/// leaf. A deeper expression is a parse error in every runtime.
pub const MAX_EXPRESSION_DEPTH: usize = 64;

/// Recursive-descent parser over a lexer's token stream (must end in EOF).
pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
    /// Nodes whose children are being parsed: each is an ancestor of what comes next.
    open: usize,
    /// Height of the node the last parse method returned.
    height: usize,
}

impl Parser {
    /// Build a parser over a token stream.
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser {
            tokens,
            current: 0,
            open: 0,
            height: 0,
        }
    }

    /// Parse the full token stream into one AST root. Errors on trailing tokens.
    pub fn parse(&mut self) -> Result<Node, ParseError> {
        let expression = self.parse_ternary_expression()?;

        if !self.is_at_end() {
            return Err(ParseError::new(format!(
                "Unexpected token after expression: {}",
                self.peek().value
            )));
        }

        Ok(expression)
    }

    // ternary = or [ "?" ternary ":" ternary ]  (right-associative)
    fn parse_ternary_expression(&mut self) -> Result<Node, ParseError> {
        let condition = self.parse_or_expression()?;

        if self.match_type(TokenType::Question) {
            let condition_height = self.height;
            self.enter()?;
            let true_value = self.parse_ternary_expression()?;
            let true_height = self.height;

            if !self.match_type(TokenType::Colon) {
                return Err(ParseError::new("Missing colon in ternary expression"));
            }

            let false_value = self.parse_ternary_expression()?;
            self.open -= 1;

            self.node(condition_height.max(true_height).max(self.height))?;
            return Ok(Node::Ternary {
                condition: Box::new(condition),
                true_value: Box::new(true_value),
                false_value: Box::new(false_value),
            });
        }

        Ok(condition)
    }

    // or = and { "||" and }
    fn parse_or_expression(&mut self) -> Result<Node, ParseError> {
        let mut left = self.parse_and_expression()?;
        let mut height = self.height;

        while self.match_type(TokenType::Or) {
            let right = self.parse_and_expression()?;
            height = self.node(height.max(self.height))?;
            left = Node::Binary {
                operator: "||".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    // and = not { "&&" not }
    fn parse_and_expression(&mut self) -> Result<Node, ParseError> {
        let mut left = self.parse_not_expression()?;
        let mut height = self.height;

        while self.match_type(TokenType::And) {
            let right = self.parse_not_expression()?;
            height = self.node(height.max(self.height))?;
            left = Node::Binary {
                operator: "&&".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    // not = "!" not | comparison
    fn parse_not_expression(&mut self) -> Result<Node, ParseError> {
        if self.match_type(TokenType::Not) {
            self.enter()?;
            let operand = self.parse_not_expression()?;
            self.open -= 1;
            self.node(self.height)?;
            return Ok(Node::Unary {
                operator: "!".to_string(),
                operand: Box::new(operand),
            });
        }

        self.parse_comparison()
    }

    // comparison = primary [ cmp_op cmp_value | in_op value_list ]
    fn parse_comparison(&mut self) -> Result<Node, ParseError> {
        let left = self.parse_primary()?;
        let left_height = self.height;

        if self.match_any(&[TokenType::In, TokenType::NotIn]) {
            let negated = self.previous().token_type == TokenType::NotIn;
            let list = self.parse_value_list()?;
            self.node(left_height)?;
            return Ok(Node::In {
                negated,
                value: Box::new(left),
                list,
            });
        }

        if self.match_any(&[
            TokenType::Eq,
            TokenType::Ne,
            TokenType::Gt,
            TokenType::Ge,
            TokenType::Lt,
            TokenType::Le,
        ]) {
            let operator = self.previous().value.clone();
            let right = self.parse_comparison_value()?;
            self.node(left_height.max(self.height))?;
            return Ok(Node::Binary {
                operator,
                left: Box::new(left),
                right: Box::new(right),
            });
        }

        Ok(left)
    }

    // value_list = "[" value { "," value } "]" | value { "," value }
    fn parse_value_list(&mut self) -> Result<Vec<Node>, ParseError> {
        let mut values = Vec::new();

        let has_brackets = self.match_type(TokenType::LBracket);

        values.push(self.parse_value_list_item()?);

        while self.match_type(TokenType::Comma) {
            values.push(self.parse_value_list_item()?);
        }

        if has_brackets && !self.match_type(TokenType::RBracket) {
            return Err(ParseError::new("Missing closing bracket in value list"));
        }

        Ok(values)
    }

    fn parse_value_list_item(&mut self) -> Result<Node, ParseError> {
        // Unquoted identifiers in an "in" list are string literals.
        if self.match_type(TokenType::Identifier) {
            return Ok(Node::Literal(LiteralValue::Str(
                self.previous().value.clone(),
            )));
        }

        if self.match_type(TokenType::Number) {
            return Ok(Node::Literal(number_literal(&self.previous().literal)));
        }

        if self.match_type(TokenType::String) {
            return Ok(Node::Literal(LiteralValue::Str(string_literal(
                &self.previous().literal,
            ))));
        }

        Err(ParseError::new(format!(
            "Invalid value in list: {}",
            self.peek().value
        )))
    }

    /// Right-hand side of a comparison. An unquoted identifier with NO following
    /// dot is a string literal (`.country == US` → 'US'); an identifier followed
    /// by a dot is a path reference, so backtrack and parse it as a path.
    fn parse_comparison_value(&mut self) -> Result<Node, ParseError> {
        if self.check(TokenType::Identifier) {
            let saved = self.current;
            self.advance(); // consume identifier

            if !self.check(TokenType::Dot) {
                self.height = 1;
                return Ok(Node::Literal(LiteralValue::Str(
                    self.previous().value.clone(),
                )));
            }

            // Followed by a dot → path reference; backtrack.
            self.current = saved;
        }

        self.parse_primary()
    }

    // primary = "(" or ")" | path | literal
    fn parse_primary(&mut self) -> Result<Node, ParseError> {
        if self.match_type(TokenType::LParen) {
            self.enter()?;
            let expression = self.parse_or_expression()?;
            self.open -= 1;
            if !self.match_type(TokenType::RParen) {
                return Err(ParseError::new("Missing closing parenthesis"));
            }
            self.node(self.height)?;
            return Ok(Node::Group(Box::new(expression)));
        }

        // A path or a literal is a leaf.
        self.height = 1;

        if self.check(TokenType::Dot)
            || self.check(TokenType::DotDot)
            || self.check(TokenType::Identifier)
        {
            return self.parse_path();
        }

        if self.match_type(TokenType::String) {
            return Ok(Node::Literal(LiteralValue::Str(string_literal(
                &self.previous().literal,
            ))));
        }

        if self.match_type(TokenType::Number) {
            return Ok(Node::Literal(number_literal(&self.previous().literal)));
        }

        if self.match_type(TokenType::Boolean) {
            let b = matches!(self.previous().literal, Literal::Bool(true));
            return Ok(Node::Literal(LiteralValue::Bool(b)));
        }

        if self.match_type(TokenType::Null) {
            return Ok(Node::Literal(LiteralValue::Null));
        }

        Err(ParseError::new(format!(
            "Unexpected token in expression: {}",
            self.peek().value
        )))
    }

    // path = relative_path | absolute_path
    fn parse_path(&mut self) -> Result<Node, ParseError> {
        let mut relative = false;
        let mut levels_up: i64 = 0;
        let mut segments: Vec<PathSegment> = Vec::new();

        if self.match_type(TokenType::DotDot) {
            relative = true;
            let dots = match self.previous().literal {
                Literal::DotCount(n) => n,
                _ => 0,
            };
            levels_up = dots - 1;
        } else if self.match_type(TokenType::Dot) {
            relative = true;
            levels_up = 0;
        }

        if self.match_type(TokenType::Identifier) {
            segments.push(PathSegment::Identifier(self.previous().value.clone()));
        } else if relative {
            return Err(ParseError::new("Missing field name after dot prefix"));
        }

        while self.match_type(TokenType::Dot) {
            if self.match_type(TokenType::Asterisk) {
                segments.push(PathSegment::Wildcard);
            } else if self.match_type(TokenType::Number) {
                let idx = match &self.previous().literal {
                    Literal::Int(n) => *n,
                    Literal::Float(f) => *f as i64,
                    _ => 0,
                };
                segments.push(PathSegment::Index(idx));
            } else if self.match_type(TokenType::Identifier) {
                segments.push(PathSegment::Identifier(self.previous().value.clone()));
            } else {
                return Err(ParseError::new(format!(
                    "Invalid path segment after dot: {}",
                    self.peek().value
                )));
            }
        }

        Ok(Node::Path {
            relative,
            levels_up,
            segments,
        })
    }

    /// Open a node whose children follow; the tree is at least one level deeper
    /// than the open nodes.
    fn enter(&mut self) -> Result<(), ParseError> {
        self.open += 1;
        if self.open >= MAX_EXPRESSION_DEPTH {
            return Err(too_deep());
        }
        Ok(())
    }

    /// Record a node whose tallest child has `child_height` as the last parsed
    /// node and return its height.
    fn node(&mut self, child_height: usize) -> Result<usize, ParseError> {
        self.height = child_height + 1;
        if self.height > MAX_EXPRESSION_DEPTH {
            return Err(too_deep());
        }
        Ok(self.height)
    }

    fn match_type(&mut self, t: TokenType) -> bool {
        if self.check(t) {
            self.advance();
            return true;
        }
        false
    }

    fn match_any(&mut self, types: &[TokenType]) -> bool {
        for &t in types {
            if self.check(t) {
                self.advance();
                return true;
            }
        }
        false
    }

    fn check(&self, t: TokenType) -> bool {
        if self.is_at_end() {
            return false;
        }
        self.peek().token_type == t
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }

    fn is_at_end(&self) -> bool {
        self.peek().token_type == TokenType::Eof
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }
}

fn too_deep() -> ParseError {
    ParseError::new(format!(
        "Expression is nested more than {MAX_EXPRESSION_DEPTH} levels deep"
    ))
}

fn number_literal(lit: &Literal) -> LiteralValue {
    match lit {
        Literal::Int(n) => LiteralValue::Int(*n),
        Literal::Float(f) => LiteralValue::Float(*f),
        _ => LiteralValue::Int(0),
    }
}

fn string_literal(lit: &Literal) -> String {
    match lit {
        Literal::Str(s) => s.clone(),
        _ => String::new(),
    }
}

File: packages/validator-php/src/Expr/Parser.php
<?php

declare(strict_types=1);

namespace CRUDUI\Validator\Expr;

/**
 * CRUDUI expression parser: Token[] -> AST (expressions.md §3/§4, JS Parser parity).
 *
 * Recursive-descent grammar (EBNF, lowest precedence first):
 *
 *   ternary    = or [ "?" ternary ":" ternary ]        (right-associative)
 *   or         = and { "||" and }
 *   and        = not { "&&" not }
 *   not        = "!" not | comparison
 *   comparison = primary [ cmp_op cmp_value | in_op value_list ]
 *   primary    = "(" or ")" | path | literal
 *   path       = [ "." | ".." ] identifier { "." (identifier | number | "*") }
 *
 * Operator precedence and ternary right-associativity come from the grammar
 * productions, not a heuristic. No eval, no split (GRAMMAR §10).
 */
final class Parser
{
    /**
     * The most nodes on a path from the root of an expression's syntax tree to a
     * leaf. A deeper expression is a parse error in every runtime.
     */
    public const MAX_EXPRESSION_DEPTH = 64;

    private int $current = 0;

    /** Nodes whose children are being parsed: each is an ancestor of what comes next. */
    private int $open = 0;

    /** Height of the node the last parse method returned. */
    private int $height = 0;

    /**
     * Build a parser over a lexer's token stream (must end in EOF).
     *
     * @param list<Token> $tokens
     */
    public function __construct(private readonly array $tokens)
    {
    }

    /**
     * Parse the full token stream into one AST root. Throws on trailing tokens.
     */
    public function parse(): Node
    {
        $expression = $this->parseTernaryExpression();

        if (!$this->isAtEnd()) {
            throw new ParseError('Unexpected token after expression: ' . $this->peek()->value);
        }

        return $expression;
    }

    // ternary = or [ "?" ternary ":" ternary ]  (right-associative)
    private function parseTernaryExpression(): Node
    {
        $condition = $this->parseOrExpression();

        if ($this->match(TokenType::QUESTION)) {
            $conditionHeight = $this->height;
            $this->enter();
            $trueValue = $this->parseTernaryExpression();
            $trueHeight = $this->height;

            if (!$this->match(TokenType::COLON)) {
                throw new ParseError('Missing colon in ternary expression');
            }

            $falseValue = $this->parseTernaryExpression();
            $this->open--;

            $this->node(max($conditionHeight, $trueHeight, $this->height));
            return new TernaryNode($condition, $trueValue, $falseValue);
        }

        return $condition;
    }

    // or = and { "||" and }
    private function parseOrExpression(): Node
    {
        $left = $this->parseAndExpression();
        $height = $this->height;

        while ($this->match(TokenType::OR)) {
            $right = $this->parseAndExpression();
            $height = $this->node(max($height, $this->height));
            $left = new BinaryNode('||', $left, $right);
        }

        return $left;
    }

    // and = not { "&&" not }
    private function parseAndExpression(): Node
    {
        $left = $this->parseNotExpression();
        $height = $this->height;

        while ($this->match(TokenType::AND)) {
            $right = $this->parseNotExpression();
            $height = $this->node(max($height, $this->height));
            $left = new BinaryNode('&&', $left, $right);
        }

        return $left;
    }

    // not = "!" not | comparison
    private function parseNotExpression(): Node
    {
        if ($this->match(TokenType::NOT)) {
            $this->enter();
            $operand = $this->parseNotExpression();
            $this->open--;
            $this->node($this->height);
            return new UnaryNode('!', $operand);
        }

        return $this->parseComparison();
    }

    // comparison = primary [ cmp_op cmp_value | in_op value_list ]
    private function parseComparison(): Node
    {
        $left = $this->parsePrimary();
        $leftHeight = $this->height;

        if ($this->match(TokenType::IN, TokenType::NOT_IN)) {
            $negated = $this->previous()->type === TokenType::NOT_IN;
            $list = $this->parseValueList();
            $this->node($leftHeight);
            return new InNode($negated, $left, $list);
        }

        if ($this->match(
            TokenType::EQ,
            TokenType::NE,
            TokenType::GT,
            TokenType::GE,
            TokenType::LT,
            TokenType::LE,
        )) {
            $operator = $this->previous()->value;
            $right = $this->parseComparisonValue();
            $this->node(max($leftHeight, $this->height));
            return new BinaryNode($operator, $left, $right);
        }

        return $left;
    }

    // value_list = "[" value { "," value } "]" | value { "," value }
    /** @return list<Node> */
    private function parseValueList(): array
    {
        $values = [];

        $hasBrackets = $this->match(TokenType::LBRACKET);

        $values[] = $this->parseValueListItem();

        while ($this->match(TokenType::COMMA)) {
            $values[] = $this->parseValueListItem();
        }

        if ($hasBrackets && !$this->match(TokenType::RBRACKET)) {
            throw new ParseError('Missing closing bracket in value list');
        }

        return $values;
    }

    private function parseValueListItem(): Node
    {
        // Unquoted identifiers in an "in" list are string literals.
        if ($this->match(TokenType::IDENTIFIER)) {
            return new LiteralNode('string', $this->previous()->value);
        }

        if ($this->match(TokenType::NUMBER)) {
            /** @var int|float $lit */
            $lit = $this->previous()->literal;
            return new LiteralNode('number', $lit);
        }

        if ($this->match(TokenType::STRING)) {
            /** @var string $lit */
            $lit = $this->previous()->literal;
            return new LiteralNode('string', $lit);
        }

        throw new ParseError('Invalid value in list: ' . $this->peek()->value);
    }

    /**
     * Right-hand side of a comparison. An unquoted identifier with NO following
     * dot is a string literal (`.country == US` → 'US'); an identifier followed
     * by a dot is a path reference, so backtrack and parse it as a path.
     */
    private function parseComparisonValue(): Node
    {
        if ($this->check(TokenType::IDENTIFIER)) {
            $saved = $this->current;
            $this->advance(); // consume identifier

            if (!$this->check(TokenType::DOT)) {
                $this->height = 1;
                return new LiteralNode('string', $this->previous()->value);
            }

            // Followed by a dot → path reference; backtrack.
            $this->current = $saved;
        }

        return $this->parsePrimary();
    }

    // primary = "(" or ")" | path | literal
    private function parsePrimary(): Node
    {
        if ($this->match(TokenType::LPAREN)) {
            $this->enter();
            $expression = $this->parseOrExpression();
            $this->open--;
            if (!$this->match(TokenType::RPAREN)) {
                throw new ParseError('Missing closing parenthesis');
            }
            $this->node($this->height);
            return new GroupNode($expression);
        }

        // A path or a literal is a leaf.
        $this->height = 1;

        if (
            $this->check(TokenType::DOT)
            || $this->check(TokenType::DOT_DOT)
            || $this->check(TokenType::IDENTIFIER)
        ) {
            return $this->parsePath();
        }

        if ($this->match(TokenType::STRING)) {
            /** @var string $lit */
            $lit = $this->previous()->literal;
            return new LiteralNode('string', $lit);
        }

        if ($this->match(TokenType::NUMBER)) {
            /** @var int|float $lit */
            $lit = $this->previous()->literal;
            return new LiteralNode('number', $lit);
        }

        if ($this->match(TokenType::BOOLEAN)) {
            /** @var bool $lit */
            $lit = $this->previous()->literal;
            return new LiteralNode('boolean', $lit);
        }

        if ($this->match(TokenType::NULL)) {
            return new LiteralNode('null', null);
        }

        throw new ParseError('Unexpected token in expression: ' . $this->peek()->value);
    }

    // path = relative_path | absolute_path
    private function parsePath(): PathNode
    {
        $relative = false;
        $levelsUp = 0;
        $segments = [];

        if ($this->match(TokenType::DOT_DOT)) {
            $relative = true;
            /** @var int $dots */
            $dots = $this->previous()->literal;
            $levelsUp = $dots - 1;
        } elseif ($this->match(TokenType::DOT)) {
            $relative = true;
            $levelsUp = 0;
        }

        if ($this->match(TokenType::IDENTIFIER)) {
            $segments[] = PathSegment::identifier($this->previous()->value);
        } elseif ($relative) {
            throw new ParseError('Missing field name after dot prefix');
        }

        while ($this->match(TokenType::DOT)) {
            if ($this->match(TokenType::ASTERISK)) {
                $segments[] = PathSegment::wildcard();
            } elseif ($this->match(TokenType::NUMBER)) {
                /** @var int|float $lit */
                $lit = $this->previous()->literal;
                $segments[] = PathSegment::index((int) $lit);
            } elseif ($this->match(TokenType::IDENTIFIER)) {
                $segments[] = PathSegment::identifier($this->previous()->value);
            } else {
                throw new ParseError('Invalid path segment after dot: ' . $this->peek()->value);
            }
        }

        return new PathNode($relative, $levelsUp, $segments);
    }

    /** Open a node whose children follow; the tree is at least one level deeper than the open nodes. */
    private function enter(): void
    {
        $this->open++;
        if ($this->open >= self::MAX_EXPRESSION_DEPTH) {
            throw self::tooDeep();
        }
    }

    /** Record a node whose tallest child has $childHeight as the last parsed node; returns its height. */
    private function node(int $childHeight): int
    {
        $this->height = $childHeight + 1;
        if ($this->height > self::MAX_EXPRESSION_DEPTH) {
            throw self::tooDeep();
        }
        return $this->height;
    }

    private static function tooDeep(): ParseError
    {
        return new ParseError('Expression is nested more than ' . self::MAX_EXPRESSION_DEPTH . ' levels deep');
    }

    private function match(TokenType ...$types): bool
    {
        foreach ($types as $type) {
            if ($this->check($type)) {
                $this->advance();
                return true;
            }
        }
        return false;
    }

    private function check(TokenType $type): bool
    {
        if ($this->isAtEnd()) {
            return false;
        }
        return $this->peek()->type === $type;
    }

    private function advance(): Token
    {
        if (!$this->isAtEnd()) {
            $this->current++;
        }
        return $this->previous();
    }

    private function isAtEnd(): bool
    {
        return $this->peek()->type === TokenType::EOF;
    }

    private function peek(): Token
    {
        return $this->tokens[$this->current];
    }

    private function previous(): Token
    {
        return $this->tokens[$this->current - 1];
    }
}
